### PR TITLE
Support building and running tests with podman

### PR DESCRIPTION
**Pull request by tmds**
This is building and tests pass on Fedora with `podman` from Fedora repos and `s2i` from https://github.com/openshift/source-to-image/releases.

Some tests are not passing with `podman` and are being skipped. I need to look into this further.

Based on this, I've created some issues for podman vs docker differences: https://github.com/containers/libpod/issues/3184, https://github.com/containers/libpod/issues/3191

Fixes https://github.com/redhat-developer/s2i-dotnetcore/issues/274

### CODE REVIEW

File: 2.1/build/test/testcommon [hunk 1]
```diff
@@ -4,21 +4,47 @@
...
 
+_S2I_PATH=$(command -v s2i)
+if [ -z "$_S2I_PATH" ]; then
+  error "s2i is not on PATH. Please install s2i or download it from https://github.com/openshift/source-to-image."
+  exit 1
+fi
+
+_PODMAN_PATH=$(command -v podman)
+_DOCKER_COMMAND=$(command -v docker)
+if [ "$_DOCKER_COMMAND" ]; then
```
Comment (omajid) [inline 2.1/build/test/testcommon]: I am a little confused here. We are determining docker vs podman by checking `command -v`. But down below, when we test for `USING_PODMAN`, we still end up using the `docker` command. It seems to me like one part is assuming `docker` command exists even when using podman, but here we are assuming the opposite.

Comment (tmds) [reply 2.1/build/test/testcommon]: When USING_PODMAN, every `docker` invocation is actually a `podman` invocation.

Comment (tmds) [reply 2.1/build/test/testcommon]: I'll change the invocations in the USING_PODMAN blocks to `podman` so it is less confusing.

Comment (omajid) [reply 2.1/build/test/testcommon]: > When USING_PODMAN, every docker invocation is actually a podman invocation.

Right, so wouldn't 

     if [ "$_DOCKER_COMMAND" ]; then

Still be true in those environments? It seems to me as those environments would then be identified as docker, not podman.

Comment (tmds) [reply 2.1/build/test/testcommon]: The script gets sourced and then it detects whether podman is there.
If the user has setup an `alias podman=docker` that won't be in effect in this script.
So this is working as intended.
Let me add a `-f` (file exist) to make the check more robust.
Is that ok for you?

Comment (tmds) [reply 2.1/build/test/testcommon]: I've improved the check based on our chat.

**Comment by tmds** [comment]
:tada: This ran successful on f30 ci for building the centos and ubi images.

On f29 and rhel8 it doesn't due to bugs/missing features in the podman stack.
I hope in the near future this will also 'just work' on rhel8 when the packages gets updated.

You can run this rootless, but then you have to update your inotify and open fd limits. Otherwise the test will fail in strange ways. Running as `sudo` is the simplest to work around those limits.

@aslicerh @omajid can you review the proposed changes?

**Comment by aslicerh** [comment]
Woo! Thanks for all this. I'll look it over later today, but I just wanted to address something real quick:

>  I hope in the near future this will also 'just work' on rhel8 when the packages gets updated.

podman is expected be updated in 8.1 to a version that supports 'podman cp', so I expect that it would.

**Comment by tmds** [comment]
> podman is expected be updated in 8.1 to a version that supports 'podman cp', so I expect that it would.

Great! The build on rhel8 got pretty far, until it hit the missing `podman cp` in `test_devmode`.

**Comment by aslicerh** [comment]
> Great! The build on rhel8 got pretty far, until it hit the missing podman cp in test_devmode.

Locally I've replaced the `podman cp` with mounting a directory on the host and manipulating the files that way and it has passed the devmode and split_build tests, just to add.

**Comment by tmds** [comment]
> Locally I've replaced the podman cp with mounting a directory on the host and manipulating the files that way and it has passed the devmode and split_build tests, just to add.

If you share the code I can put it in. Unless you don't mind waiting for rhel8.1?
My main interest is this works on f30 because that is what I use for development.

**Comment by tmds** [comment]
If there are no review comments, I will copy over the changes to the 2.2 version and then merge this after a successful run on f30 CI.

**Comment by aslicerh** [comment]
This all looks good to me. Thanks for all the hard work!

Also, I like your solution to the libpod issue much better than the terrible and hacky one I had for my own tests!

**Comment by tmds** [comment]
Thanks Andrew, I'm sure your workarounds weren't too bad :)

**Comment by tmds** [comment]
Our f30 CI vm is broken. I ran the tests locally and they passed.

**Pull request merged**